turbo: updating all docs to current state

thlorenz opened the pull request:
## Updating to latest v8

Updated all the documents to reflect changes in v8, mainly the compiler chain as well as the garbage collector.

Numerous documents focus on one topic specifically and I include a quick summary + link to the document below as well as the diff. I did that b/c I feel looking at the diff makes things much harder to read than when looking at the original file, but comments would make sense in the diff.
I'm hoping that the experts from the v8 team and community can review a document that most matches their expertise and point out any information that is not quite correct so I can correct it.

**Please note that all documents inside the `./crankshaft` folder are the old versions and don't need to be reviewed. Please review the ones at the top level of this repo (and linked below) instead.**

Obviously pull requests are also very welcome :)

### Compiler

The [v8 compiler](https://github.com/thlorenz/v8-perf/blob/turbo-updates/compiler.md) | [diff](https://github.com/thlorenz/v8-perf/compare/turbo-updates?expand=1#diff-b49978087b88c89a5fa214b85931590b) document outlines the v8 compiler pipeline including the
Ignition Interpreter and TurboFan optimizing compiler. It explains how information about your
code is executed to allow optimizations, how and when deoptimizations occur and how features
like the CodeStubAssembler allowed reduce performance bottlenecks found in the [older
pipeline](https://github.com/thlorenz/v8-perf/blob/turbo-updates/crankshaft/compiler.md).

### Language Features

The [language features](https://github.com/thlorenz/v8-perf/blob/turbo-updates/language-features.md) | [diff](https://github.com/thlorenz/v8-perf/compare/turbo-updates?expand=1#diff-760a1ebc2a62f1b12ce76c6947cd4390) document lists JavaScript language features and
provides info with regard to their performance mainly to provide assurance that performance of
most features is no longer an issue as it was with the previous compiler pipeline.

### Garbage Collector

The [v8 garbage collector](https://github.com/thlorenz/v8-perf/blob/turbo-updates/gc.md) | [diff](https://github.com/thlorenz/v8-perf/compare/turbo-updates?expand=1#diff-dd25f7f05abfbcd228db92bb870d89b6) document talks about how memory is organized on the v8 heap,
how garbage collection is performed and how it was parallelized as much as possible to avoid
pausing the main thread more than necessary.

### Memory Profiling

The [memory profiling](https://github.com/thlorenz/v8-perf/blob/turbo-updates/memory-profiling.md) | [diff](https://github.com/thlorenz/v8-perf/compare/turbo-updates?expand=1#diff-e7fe1dc69ba71e2ee2721d647d59b47d) document explains how JavaScript objects are
referenced to form a tree of nodes which the garbage collector uses to determine _collectable_
objects. It also outlines numerous techniques to profile memory leaks and allocations.

### Inspection and Performance Profiling

Inside the [inspection](https://github.com/thlorenz/v8-perf/blob/turbo-updates/inspection.md) | [diff](https://github.com/thlorenz/v8-perf/compare/turbo-updates?expand=1#diff-1ab137d931a59172f302cba93e8e965a) document you will find techniques that allow you to
profile your Node.js or web app, how to produce flamegraphs and what flags and tools are
available to gain an insight into operations of v8 itself.

### Snapshots and Code Caching

[This document](https://github.com/thlorenz/v8-perf/blob/turbo-updates/snapshots+code-caching.md) | [diff](https://github.com/thlorenz/v8-perf/compare/turbo-updates?expand=1#diff-4c875c24d36e23ba24b060d46e9f6067) includes information as to how v8 uses caching
techniques in order to avoid recompiling scripts during initialization and thus achieve faster
startup times.

### Runtime Functions

The [runtime functions](https://github.com/thlorenz/v8-perf/blob/turbo-updates/runtime-functions.md) | [diff](https://github.com/thlorenz/v8-perf/compare/turbo-updates?expand=1#diff-82eb773b939d6a4a0b9c673cb78c2689) document gives a quick intro into C++ functions
accessible from JavaScript that can be used to provide information of the v8 engine as well as
direct it to take a specific action like optimize a function on next call.

cc @mraleph @bmeurer @mcollina @mathiasbynens @fhinkel @ofrobots

I realize these are by no means all people knowledgable about these topics, so please feel free to add more via a comment.

I very much appreciate anyone taking the time to help make this a valuable resource with correct and up to date information.

bmeurer commented on crankshaft/compiler.md:
```diff
@@ -0,0 +1,281 @@
+
+## Components
+
+### Base Compiler 
```
The *Base Compiler* is the Ignition interpreter.

thlorenz replied:
Sorry should have made this clear, but all documents inside the `./crankshaft` folder are from the previous version and kept there only in case people want to compare or need to see how things used to work.

I'm mainly interested in reviews of the new content at the root of this repo (basically the ones I linked in the PR message).

However any of your comments that also apply to newer content I will apply there.

thlorenz replied:
So in this case that'd be the [compiler.md](https://github.com/thlorenz/v8-perf/blob/turbo-updates/compiler.md)

bmeurer commented on crankshaft/compiler.md:
```diff
@@ -0,0 +1,281 @@
+### Base Compiler 
+
+- is used for all code initially
+- generates code quickly without heavy optimizations
```
More specifically: Generates bytecode

thlorenz replied:
Not present in updated compiler.md anymore.

bmeurer commented on crankshaft/compiler.md:
```diff
@@ -0,0 +1,281 @@
+
+- recompiles and optimizes hot code identified by the runtime profiler
+- uses static single assignment form to perform optimizations
+  - loop-invariant code motion
```
We don't really perform LICM in TurboFan. Instead we do loop peeling and redundancy elimination.

thlorenz replied:
Not present in updated compiler.md anymore.

bmeurer commented on crankshaft/compiler.md:
```diff
@@ -0,0 +1,281 @@
+[watch](http://youtu.be/VhpdsjBUS3g?t=18m53s)
+[watch](http://youtu.be/UJPdhx5zTaw?t=26m30s)
+
+### Full Compiler
```
As said, this is Ignition now. Fullcodegen doesn't exist anymore.

thlorenz replied:
Not present in updated compiler.md anymore.

mathiasbynens commented on README.md:
```diff
@@ -1,55 +1,95 @@
+  - [Inspection and Performance Profiling](#inspection-and-performance-profiling)
+  - [Snapshots and Code Caching](#snapshots-and-code-caching)
+  - [Runtime Functions](#runtime-functions)
+- [v8 source and documentation](#v8-source-and-documentation)
```
Nit: The project name is “V8” with a capital V.

mathiasbynens commented on README.md:
```diff
@@ -1,55 +1,95 @@
+The [v8 compiler](compiler.md) document outlines the v8 compiler pipeline including the
+Ignition Interpreter and TurboFan optimizing compiler. It explains how information about your
+code is executed to allow optimizations, how and when deoptimizations occur and how features
+like the CodeStubAssembler allowed reduce performance bottlenecks found in the [older
```
reducing

mathiasbynens commented on crankshaft/compiler.md:
```diff
@@ -0,0 +1,281 @@
+
+- monitors the running system and identifies hot code
+
+### Optimizing Compiler 
```
Let’s add `(TurboFan)` to the heading

thlorenz replied:
As stated above please don't look at the docs in the `./crankshaft` folder.
That heading doesn't exist in the new compiler.md (at root level) anymore

mathiasbynens commented on crankshaft/compiler.md:
```diff
@@ -0,0 +1,281 @@
+- uses static single assignment form to perform optimizations
+  - loop-invariant code motion
+  - linear-scan register allocation
+  - inlining.
```
Nit: remove `.`

thlorenz replied:
ditto

mathiasbynens commented on crankshaft/compiler.md:
```diff
@@ -0,0 +1,281 @@
+## Optimized Code vs. Inline Caches and Unoptimized Code
+
+[watch](http://youtu.be/VhpdsjBUS3g?t=18m53s)
+[watch](http://youtu.be/UJPdhx5zTaw?t=26m30s)
```
Can haz more descriptive link text?

thlorenz replied:
Comment is on old content, but I carried the way of doing things into the new one as well, so if it's a problem I'd need to fix it there.

The idea is that I have resources with descriptive title at the end of sections or documents.
The little `[watch]` and `[read]` links are always in subsections and targeting exactly the topic discussed, as in if you need more info about what you're just reading, click on it.
They are meant to be small in order not to break the reading flow.

I'm open to suggestions though on how to improve on this.

thlorenz replied:
Would the combination of a short link with italicized description work, i.e. like [here](https://github.com/thlorenz/v8-perf/blob/turbo-updates/data-types.md#object-properties)?

That could be compromise of keeping those links out of your eye but providing an idea of what awaits you when you follow it.

mathiasbynens commented on crankshaft/compiler.md:
```diff
@@ -0,0 +1,281 @@
+- 1. validate type assumptions (are hidden classes as expected)
+- 2. do work
+- change at runtime via backpatching as more types are discovered to generate new ICs
+  [watch](http://youtu.be/UJPdhx5zTaw?t=28m44s) | [slide](http://v8-io12.appspot.com/index.html#56)
```
Let’s use HTTPS for all YouTube and appspot links

mathiasbynens commented on crankshaft/compiler.md:
```diff
@@ -0,0 +1,281 @@
+- optimizations are speculative and assumptions are made
+- if assumption is violated
+  - function deoptimized
+  - execution resumes in full compiler code
```
full compiler → interpreter (Ignition)

thlorenz replied:
See new compiler.md.

mathiasbynens commented on crankshaft/compiler.md:
```diff
@@ -0,0 +1,281 @@
+  - if assumptions are violated again, deoptimized again and start over 
+- too many deoptimizations cause function to be sent to *deoptimization hell*
+  - considered not optimizable and no optimization is **ever** attempted again
+- certain constructs like `try/catch` are considered not optimizable and functions containing it go straight to
```
This was true in Crankshaft but it’s no longer true with the new pipeline! Please remove this.

thlorenz replied:
Yes, that's why it's in the document inside the `./crankshaft/` folder.
I'm really sorry I didn't make this more clear :(
I assumed reviewers could just follow the links in the PR description.

I added a bold note to the description to avoid this from happening to other reviewers.

mathiasbynens commented on crankshaft/compiler.md:
```diff
@@ -0,0 +1,281 @@
+[watch](http://youtu.be/UJPdhx5zTaw?t=15m35s) | [slide](http://v8-io12.appspot.com/index.html#34)
+
+- v8 passes around 32bit numbers to represent all values
+- bottom bit reserved as tag to signify if value is a SMI (small integer) or a pointer to an object
```
`s/SMI/Smi/g`

mathiasbynens commented on crankshaft/compiler.md:
```diff
@@ -0,0 +1,281 @@
+
+or
+
+| 31-bit-signed integer (SMI) | 0 |
```
It’s actually Signed31 on 32-bit, Signed32 on 64-bit IIRC

mathiasbynens commented on crankshaft/compiler.md:
```diff
@@ -0,0 +1,281 @@
+
+### Double Array Unboxing
+
+[watch](http://youtu.be/UJPdhx5zTaw?t=20m20s) | [slide](http://v8-io12.appspot.com/index.html#45)
```
Alternate, more recent presentation: https://www.youtube.com/watch?v=m9cTaYI95Zc Slides are linked in the description

thlorenz replied:
Video and post were already linked in resources of the updated data-types.md.
I added the slides as well now.

However I couldn't see a part in the video that focuses particularly on double array unboxing the way that the linked (out of date) video + slide does. Did I miss something and if so please point me to the slide or time in the up to date video.

davidmarkclements commented on crankshaft/compiler.md:
```diff
@@ -0,0 +1,281 @@
+
+### Considerations
+
+- prefer SMIs for numeric values whenever possible
```
also, if a non-SMI is used in a calculation the result will be non-SMI even if the result is in bounds (and then even if it's in bounds and it's considered non-SMI it will continue to make results non-SMI)

```js
var n = 2147483647
%IsSmi(n) // true
n++
%IsSmi(n) // false
n--
%IsSmi(n) // false
%IsSmi(n - 1) // false
```

thlorenz replied:
That is an interesting edge-case, however I feel that it adds more confusion to include that in the document (there are lots of things I left out to keep these consumable and understandable).
Also in a way that's explained as part of [double unboxing](https://github.com/thlorenz/v8-perf/blob/turbo-updates/data-types.md#double-array-unboxing) at least for Arrays.

Maybe we could add a note that this applies to individual variables as well? Do you have a suggestion to succinctly state that there?

* * *

WRT https://github.com/thlorenz/v8-perf/pull/14#discussion_r181133693 (again github doesn't let me comment).
Yes I'd think it's bit early to talk about BigInt. Also I didn't find any v8 dev publications about that topic.

thlorenz commented:
Thanks for the reviews so far @bmeurer and @mathiasbynens.

I'm terribly sorry for the confusion as to what's the new content to be reviewed.
I addressed the issues that did affect the new content and hope you can find the time to review those new documents as well so I can fix any issues there as well.

As stated anything in the `./crankshaft` folder needs no review as that's old content that I keep there just for future reference.
However lots of content was added in the top level of this repo and I linked them directly from the PR message.

thlorenz commented on crankshaft/compiler.md:
```diff
@@ -0,0 +1,281 @@
+
+- operations are monomorphic if hidden classes of arguments are **always** same
+- all others are polymorphic at best and megamorphic at worst
+- monomorphic operations are easier optimized
```
Object shapes are explained inside data-types.md around [here](https://github.com/thlorenz/v8-perf/blob/turbo-updates/data-types.md#objects) and below map transitions are explained as well, however if we need more content I can add that either to the new compiler.md or inside the data-types doc.

thlorenz commented on crankshaft/compiler.md:
```diff
@@ -0,0 +1,281 @@
+
+#### Considerations
+
+- prefer monomorphic over polymorphic functions wherever possible
```
That section doesn't exist in the new doc anymore, but maybe we can add it somewhere?
At this point I'm just [recommending](https://github.com/thlorenz/v8-perf/blob/turbo-updates/compiler.md#recommendations) to write _good_ JavaScript and not worry about workarounds.

davidmarkclements replied:
Yeah I think that's best in most cases .. although polymorphic will always be slower and it might be worth highlighting still

thlorenz commented on crankshaft/compiler.md:
```diff
@@ -0,0 +1,281 @@
+
+[watch](http://youtu.be/UJPdhx5zTaw?t=33m12s) | [slide](http://v8-io12.appspot.com/index.html#65)
+
+- if function executes a lot it becomes **hot**
```
not part of new content anymore

thlorenz commented on crankshaft/compiler.md:
```diff
@@ -0,0 +1,281 @@
+  this.y = y;
+}
+
+var p = new Point(1, 2); // => hidden Point class created
```
I agree that we should point that out (I'm not doing that in new doc either).
WRT an example I believe this one is more straightforward to understand, but I'll add a note WRT object literals.

Somehow github doesn't allow me to comment on previous comment.
I am aware of deopt cases that turned out to be v8 bugs, but the consensus is now to tell the V8 team about it instead of trying to workaround.

In the new compiler.md document I [do talk about deptimization](https://github.com/thlorenz/v8-perf/blob/turbo-updates/compiler.md#deoptimization), one [deopt loop bug](https://github.com/thlorenz/v8-perf/blob/turbo-updates/compiler.md#deoptimization-loop) that's been fixed and [an example](https://github.com/thlorenz/v8-perf/blob/turbo-updates/compiler.md#class-definitions-inside-functions) in which `usePoint` will essentially go to deoptimization hell.

If you have other examples we should add I can do that around that section.

thlorenz replied:
Addressed in 34a8f45f20f2d467f332a654ad10f17126e0feb9

thlorenz commented on crankshaft/compiler.md:
```diff
@@ -0,0 +1,281 @@
+##### Considerations
+
+- avoid hidden class changes
+- initialize all members in **constructor function** and **in the same order**
```
Yes they do, but it is easier to guarantee same properties + same order if we use a `constructor` as then we have one place in which properties are assigned.

It does apply to classes as well and I will point that out more clearly.

thlorenz replied:
Addressed in 34a8f45f20f2d467f332a654ad10f17126e0feb9

thlorenz commented:
@mathiasbynens I addressed the SMIs issue in 62580518f3dc6e94c4257ae45e8128d75fab697a.
(github doesn't let me comment on your comment .. no idea why)

ofrobots commented on memory-profiling.md:
```diff
@@ -123,126 +131,112 @@
 
-The usefulness of the information presented in the below views depends on how you authored your code. Here are a few
-rules to make your code more debuggable.
+![memory-hog-allocation-timeline](assets/memory-allocation-timeline.gif)
```
Wrong gif?

thlorenz replied:
Fixed in a0a098d2ad4e72edd446729a5b469fb2c43b3c33

thlorenz commented on memory-profiling.md:
```diff
@@ -123,126 +131,112 @@
+- stop profiling via the red circle on the upper left and select _Chart_ from the dropdown on
+  the left
+- you will see a function execution stack with the functions that allocated the most memory or
+  had children that executed lots of memory being the widest
```
Added in a0a098d2ad4e72edd446729a5b469fb2c43b3c33

thlorenz commented on crankshaft/gc.md:
```diff
@@ -0,0 +1,350 @@
+  
+### Old Data Space
+
+- contains *raw data* objects -- **no pointers**
```
Found the origin of this info. It's in the _Heap Organization_ section of [this rather old post](http://jayconrod.com/posts/55/a-tour-of-v8-garbage-collection).

ulan replied:
We don't have data space anymore. Data and pointer spaces merged some time ago.

thlorenz replied:
OK, thanks I'll update the doc. Is there a writeup somewhere that I could reference to get this correct?

ulan replied:
I don't know if it was written up anywhere. The source code is probably the best reference:
https://cs.chromium.org/chromium/src/v8/src/globals.h?rcl=8457e810efd34381448d51d93f50079cf1f6a812&l=525

There ongoing work on read-only (RO) space.

thlorenz replied:
Thanks, I'll read through that and update the related notes tomorrow.

thlorenz replied:
Addressed in 9d79c90

ulan commented on gc.md:
```diff
@@ -189,53 +226,133 @@ ToSpace starts as unallocated memory.
-[barrier](http://www.memorymanagement.org/glossary/b.html#term-barrier-1) | [write
-barrier](http://www.memorymanagement.org/glossary/w.html#term-write-barrier) | [read
-barrier](http://www.memorymanagement.org/glossary/r.html#term-read-barrier)
+- every allocation brings us closer to GC pause
```
By now Mark-Compact GC is mostly concurrent, Scavenger is parallel, so GC pauses should be small: 1ms - 10ms on average.

The pause Mark-Compact GC depends on how many objects are alive on the heap and how fragmented the heap is. If you observe large pauses (>100ms) I would be interested to take a look at the workload.

thlorenz replied:
OK I'll add that info in, thanks.
However the fact still remains that you want to avoid allocating/collecting more objects than necessary even though the penalty isn't as bad anymore right?

ulan replied:
If it possible without sacrificing code quality then yes. Otherwise, it depends :)
For example, temporary short-living objects that are allocated in the young generation and die on the next Scavenge have almost zero cost. This is because GC does not visit dead objects, it visits only live objects.

thlorenz replied:
Also I'm bit unclear WRT to allocating + immediately filling a large Array and performance.
It used to be that you'd want to pre-alloc as long as it had `<100K` items, in the below example function _one_ performed better than _two_.

Is this still true?

```js
function one(items) {
  const result = new Array(items.length)
  for (var i = 0; i < items.length; i++) {
    result[i] = items[i] * 10
  }
  return result
}

function two(items) {
  const result = []
  for (const item of items) {
    result.push(item * 10)
  }
  return result
}
```

thlorenz replied:
Included the following detail:

> temporary short-living objects that are allocated in the young generation and die on the next Scavenge have almost zero cost. This is because GC does not visit dead objects, it visits only live objects.

here efe68a1.

This was a bit tricky as I'm trying to keep those considerations short.
The details of scavenging and why objects surviving a Scavenge are much more expensive (copying) is outlined in the sections just above, so I think it should be covered.

ulan commented on gc.md:
```diff
@@ -189,53 +226,133 @@ ToSpace starts as unallocated memory.
+- every allocation brings us closer to GC pause
+- even though as many steps of collection are performed in parallel, **every collection pauses our
+  app**
+- try to pre-alloc as much as possible ahead of time
```
Pre-allocation as much as possible is in general not a good idea :)
GC complexity depends on the live set. Pre-allocation artificially increases the live set.

thlorenz replied:
I was talking about preallocating things that you'll use a lot, but they don't change, i.e.:

```js
// msg allocated one time, never needs garbage collecting
const msg = 'User not found'

function onuserNotFound() {
  emit(msg)
}
```

Isn't that more efficient than:

```js
function onuserNotFound() {
  // message allocated for each invocation and will be collected over and over
  emit('User not found')
}
```

thlorenz replied:
Also I'm bit unclear WRT to (pre)allocating + immediately filling a large Array and performance.
It used to be that you'd want to pre-alloc as long as it had `<100K` items, in the below example function _one_ performed better than _two_.

Is this still true?

```js
function one(items) {
  const result = new Array(items.length)
  for (var i = 0; i < items.length; i++) {
    result[i] = items[i] * 10
  }
  return result
}

function two(items) {
  const result = []
  for (const item of items) {
    result.push(item * 10)
  }
  return result
}
```

thlorenz replied:
Tried to address in 9d79c90, but still would like clarifications about my questions here, so I can add that info.

ulan replied:
It depends. Preallocating immutable objects used in a hot loop is probably worthwhile.

Preallocating mutable object that refers to other objects or closures increases chances of memory leaks because now the developer has to manually delete unused references. Memory leaks increase memory usage and GC pauses.

Also a preallocated object slows down all subsequent GCs because each GC will have to visit it.

bmeurer replied:
Yes, that's correct. Not sure about the exact limits. I think @mathiasbynens recently checked that

ulan commented on gc.md:
```diff
@@ -1,65 +1,78 @@
+  - [Old Data Space](#old-data-space)
+  - [Large Object Space](#large-object-space)
+  - [Code Space](#code-space)
+  - [Cell Space, Property Cell Space, Map Space](#cell-space-property-cell-space-map-space)
```
Cell and Property spaces were merged to Old space.

thlorenz replied:
Addressed in 9d79c90

ulan commented on gc.md:
```diff
@@ -1,65 +1,78 @@
 ## Goals, Techniques
 
 - ensures fast object allocation, short garbage collection pauses and no memory fragmentation
 - **stop-the-world**,
```
Is this about Mark-Compact GC or Scavenger?

Mark-Compact GC is not stop-the-world. It was incremental since 2011. Recently it became mostly concurrent (blog post is coming).

thlorenz replied:
This came from the description on the [V8 wiki](https://github.com/v8/v8/wiki/Design%20Elements#efficient-garbage-collection).

But I'd assume this is more important regarding scavenging since that happens much more often and usually is the reason for perf bottlenecks due to GC.
From what I understand it's mostly parallel now as well, but there are still steps for which it has to _stop the world_ to ensure nothing is allocated on the heap while it's performing those steps.

Is that correct and sufficient to call it _stop the world_?
If so I'll just clarify that this mostly applies to the scavenger.

thlorenz replied:
Tried to address in 9d79c90.
I tried to make it more obvious in the Goals section which parts actually stop-the-world by adding a few more notes and linking to parts of the document where I explain incremental mark-compact GC.

ulan replied:
Yep, the scanveger is a stop-the-world parallel GC.

Thanks for the pointer. The V8 wiki needs update.

ulan commented on gc.md:
```diff
@@ -1,65 +1,78 @@
 
 - ensures fast object allocation, short garbage collection pauses and no memory fragmentation
 - **stop-the-world**,
   [generational](http://www.memorymanagement.org/glossary/g.html#term-generational-garbage-collection) accurate garbage collector
```
accurate => precise

thlorenz replied:
Addressed in 9d79c90

ulan commented on gc.md:
```diff
@@ -72,17 +85,37 @@ everything else is garbage*
 - objects pointed to by local variables are roots (stack is scanned for roots)
 - DOM elements are roots (may be weakly referenced)
```
DOM objects now participate in marking similar to JS objects: https://v8project.blogspot.de/2018/03/tracing-js-dom.html

thlorenz replied:
Addressed in be7aef2

ulan commented on gc.md:
```diff
@@ -72,17 +85,37 @@ everything else is garbage*
-[watch](http://youtu.be/VhpdsjBUS3g?t=11m24s)
+[watch](https://youtu.be/VhpdsjBUS3g?t=11m24s)
 
 - object heap segmented into two parts (well kind of, [see below](#heap-organization-in-detail))
```
Heap consists of young generation and old generation.
Young generation consists of new space.
Old generation consists of old space, map space, code space, large object space.

thlorenz replied:
OK I'll clarify that.

thlorenz replied:
Addressed in 9d79c90

hashseed commented on snapshots+code-caching.md:
```diff
@@ -0,0 +1,84 @@
+  - later cache data can be attached to the source object and passed
+    `v8::ScriptCompiler::kConsumeCodeCache` as an option to cause V8 to bypass compileing the
+    code and deserialize the provided cache data instead
+- V8 6.6 caches top level code as well as code generated _after_ script's top-level execution,
```
Note that kProduceCodeCache is being deprecated in favor of this new API (ScriptCompiler::GetCodeCache)

thlorenz replied:
OK I'll mention that ... also keep in mind that this is meant to be a living document (which is why I made it a github repo instead of a blog).
So please feel free to file issues/PRs when things change in the future so we can get that updated. :)
For now I'll add a note about the coming API change.

thlorenz replied:
Addressed in b11e195

hashseed commented on snapshots+code-caching.md:
```diff
@@ -0,0 +1,84 @@
+Since Chrome embeds V8 it can make use of Code Caching and does so as follows.
+
+- cold load: page loaded for the first time and thus no cached data is available
+- warm load: page loaded before and caches compiled code along with the script file in disk
```
Current heurisric also includes that the last load needs to be within the last 72 hours.

thlorenz replied:
Wow, I had no idea about that part .. I'll add that.
Is there a resource somewhere about that detail? Otherwise I'll try to find something related in the v8 code.

hashseed replied:
This is actually [here](https://cs.chromium.org/chromium/src/third_party/blink/renderer/bindings/core/v8/v8_script_runner.cc?l=269&rcl=c59618d0f92b57e4dcfb903f3c99bb0574eac340). We also require the script source to be larger than 1k.

thlorenz replied:
Addresed in b11e195

thlorenz commented on snapshots+code-caching.md:
```diff
@@ -0,0 +1,84 @@
+  execute each bytecode within Ignition interpreter
+- enabled in V8 v6.4 resulting in average V8's heap size savings of 540 KB
+
+### Resources
```
I do think it's still relevant and I'll add a note, thanks.

thlorenz replied:
From that document it's not clear when this was added to Node.js.
Also I remember reading something by @hashseed about problems regarding hash keys not being random and thus opening attack vectors due to snapshotting.

So which version of Node.js is the first one to use V8 snapshots and is it still being used?

thlorenz replied:
For now I at least added a link to the resource ad2e331

thlorenz commented on gc.md:
```diff
@@ -125,17 +158,16 @@ everything else is garbage*
 - each space divided into set of pages
 - page is **contiguous** chunk of memory allocated via `mmap`
-- page is 1MB in size and 1MB aligned 
+- page is 1MB in size and 1MB aligned
```
Addressed in 9d79c90

sigurdschneider commented on compiler.md:
```diff
@@ -1,155 +1,483 @@
+- low memory vs. max optimization
+- Ignition Interpreter allows to run code with some amount of optimization very quickly and has
+  very low memory footprint
+- TurboFan makes functions that run a lot fast, sacrificing some memory in the process
```
TurboFan is able to generates machine code for all JavaScript code, but machine code takes up considerably more space than Iginition bytecode. This is in contrast to Crankshaft, which could not generate optimized code for the full language (i.e. try/catch could not be optimized).

thlorenz replied:
The fact that optimized code takes up more memory is implied with the list item as we state that memory is sacrificed, while stating in the item just above that Ignition has low memory footprint.
However I'm open to add more details somewhere as to why if you think that's important.

The supporting the full language aspect is addressed in [this document](https://github.com/thlorenz/v8-perf/blob/turbo-updates/language-features.md) where I point out improvements feature by feature. I also included [a section in the compiler document](https://github.com/thlorenz/v8-perf/blob/turbo-updates/compiler.md#new-language-features) to add details.

sigurdschneider commented on compiler.md:
```diff
@@ -1,155 +1,483 @@
+  very low memory footprint
+- TurboFan makes functions that run a lot fast, sacrificing some memory in the process
+- designed to support entire JavaScript language and make it possible to quickly add new
+  features and to optimize them fast and incrementally
```
The new design with Ignition and TurboFan allows to integrate new language features very quickly and develop powerful optimizations incrementally.

thlorenz replied:
This is described in detail [here](https://github.com/thlorenz/v8-perf/blob/turbo-updates/compiler.md#new-language-features), not sure what we should add, but open to suggestions.

sigurdschneider commented on compiler.md:
```diff
@@ -1,155 +1,483 @@
+2. Generate bytecode from that AST
+3. Turn bytecode into sequence of bytecodes by the BytecodeGenerator, which is part of the [Ignition Interpreter](https://v8project.blogspot.com/2016/08/firing-up-ignition-interpreter.html)
+  - sequences are divided on a per function basis
+4. Execute bytecode sequences via Ignition and collect feedback via inline caches
```
(e.g. "x" is a variable that has always been holding a string)

thlorenz replied:
This is a summary of the basic steps.
The details are explained in [this section](https://github.com/thlorenz/v8-perf/blob/turbo-updates/compiler.md#collecting-feedback-via-ics) including a feedback lattice.

sigurdschneider commented on compiler.md:
```diff
@@ -1,155 +1,483 @@
+  - feedback used by Ignition itself to speed up subsequent interpretation of the bytecode
+  - feedback used for speculative optimization by TurboFan when code is optimized
+5. _Speculatively_ optimize and compile bytecode using collected feedback to generate optimized machine code
+  for the current architecture
```
Turbofan compiles and optimizes Ignition's bytecode to machine code for the current architecture, thereby using the collected feedback to make _speculative_ assumptions (i.e. TurboFan bakes in the assumption that a certain variable always holds a string).

thlorenz replied:
Again, this is just the summary.
Details on speculative optimization are found [here](https://github.com/thlorenz/v8-perf/blob/turbo-updates/compiler.md#speculative-optimization) including a code sample.

sigurdschneider commented on compiler.md:
```diff
@@ -1,155 +1,483 @@
-  optimized code turn out to be too optimistic
+1. Parse JavaScript into an [AST (abstract syntax tree)](https://en.wikipedia.org/wiki/Abstract_syntax_tree)
+2. Generate bytecode from that AST
+3. Turn bytecode into sequence of bytecodes by the BytecodeGenerator, which is part of the [Ignition Interpreter](https://v8project.blogspot.com/2016/08/firing-up-ignition-interpreter.html)
```
I think it should just say that Ignition generates the bytecode and interprets it in an /indirectly threaded/ fashion. The link to the blog pos is useful.

thlorenz replied:
Would like to keep the summary small. Could you provide a way to add this as a detail somewhere in the Ignition section?
I'm also not sure what _indirectly threaded fashion_ means in this context. The blog post I'm linking doesn't mention that exactly.
Could you elaborate and/or add a link so I can read up on this?

sigurdschneider commented on compiler.md:
```diff
@@ -1,155 +1,483 @@
+- previous V8 implementations suffered from _performance cliffs_
+  - optimized code ran super fast (focus on peak performance case)
+  - baseline performance was much lower
+  - as a result one feature in your code that caused deoptimization would affect your app's
```
The word "deoptimization" in the Crankshaft is slang for disabled optimization, which means that the JavaScript code contained a feature not supported for optimization by Crankshaft (think try/catch). Deoptimization is technically something different: it just means that one of the /speculative/ assumptions the compiler made turned out to be wrong at runtime (i.e. x was assumed to hold a string, now it holds an object), and we have to go back to the baseline (nowadays Ignition, in the old days fullcodegen).

So let's get rid of "caused deoptimization" here and just write "caused your code to fall of the fast-path"

thlorenz replied:
Thanks, changing it to _that prevented it's optimization_ which is a bit more explicit IMO

thlorenz replied:
Addressed in fe82604

sigurdschneider commented on compiler.md:
```diff
@@ -1,155 +1,483 @@
+    performance dramatically, i.e. 100x difference
+- TurboFan improves this as
+  - widens fast path to ensure that optimized code is more flexible and can accept more types
+    of arguments
```
TurboFan handles the full JavaScript language. There is no "disabled optimization" (see comment a above) anymore. How aggressively TurboFan optimizes may vary.

thlorenz replied:
This is referring to types of arguments a function can handle while still being optimized.
It is not referring to JavaScript features.
I tried to explain in detail [here](https://github.com/thlorenz/v8-perf/blob/turbo-updates/compiler.md#new-language-features) that all language features are now supported as well as address improvements one by one [here](https://github.com/thlorenz/v8-perf/blob/turbo-updates/language-features.md), but I'm open to add more info there.

Just LMK what you feel is missing there.

sigurdschneider commented on compiler.md:
```diff
@@ -1,155 +1,483 @@
+- TurboFan improves this as
+  - widens fast path to ensure that optimized code is more flexible and can accept more types
+    of arguments
+  - reduces code memory overhead by reusing code generation parts of TurboFan to build Ignition
```
e.g. certain builtins (e.g. Array.p.slice) are called from both, Ignition and TurboFan code.

sigurdschneider commented on compiler.md:
```diff
@@ -1,155 +1,483 @@
+
+- no longer relying on optimizing compiler for _sufficiently_ fast code
+- thus improved baseline performance allows delaying optimization until more feedback is collected
+- leads to less time and resources spent optimizing
```
This also allows us to not optimize code that turns out to be infrequently executed.

thlorenz replied:
Addressed in fe82604

sigurdschneider commented on compiler.md:
```diff
@@ -1,155 +1,483 @@
+- last computed value of each bytecode is kept in special _accumulator_ register minimizing
+  load/store operations (from/to explicit registers)
+- current stack frame is identified by stack pointer
+- program counter points to currently executed instruction in the bytecode
```
- each bytecode handler tail-calls into the next bytecode handler (indirectly threaded interpreter)

thlorenz replied:
Addressed in fe82604

thlorenz commented on compiler.md:
```diff
@@ -1,155 +1,483 @@
+- fast startup vs. peak performance
+- low memory vs. max optimization
+- Ignition Interpreter allows to run code with some amount of optimization very quickly and has
+  very low memory footprint
```
Here I'm just summarizing the goals/advantages.
Details are outlined [here](https://github.com/thlorenz/v8-perf/blob/turbo-updates/compiler.md#advantages-over-old-pipeline)
> bytecode is 25-50% the size of equivalent baseline machine code, [here](https://github.com/thlorenz/v8-perf/blob/turbo-updates/compiler.md#startup-time-improved)
> bytecode smaller and faster to generate than machine code (crankshaft)
and [here](https://github.com/thlorenz/v8-perf/blob/turbo-updates/compiler.md#ignition-interpreter) among other places.
> this results in highly optimized and small interpreter code

Hope that suffices, if more details are needed we can add them in one of the details section.

thlorenz commented on compiler.md:
Addressed in fe82604
